### DISCUSSION TITLE
Store output into a custom variable (was: New Variable {{statusbar:get}})

### DISCUSSION THREAD

**Discussion by FelipeRearden** [Taitava/obsidian-shellcommands#127] [Ideas - done & released]
Hello @Taitava !!!!!

This is something that is in my mind since you launch the output statusbar .

The main idea is a variable to get the content that was previously inserted in the statusbar by the output statusbar

it's a second clipboard that I can temporary store text than use the new variable to get the content from the status bar and use in the note with our fantastics outputs (caret positions or clipboard)

## My Workflow

Many times when writing about one topic I have to navigate between many notes and stop to paste the content in the target note. 

My idea is to use this new variable to help to store more content when researching:
- {{selection}} from noteA inside {{statusbar}}
- {{clipboard}} from noteB
- {{statusbar:get}} and {{clipboard}}  pasted in noteC

SC#1 - echo {{selection}} [output:statusbar]
 
SC#2 - echo {{statusbar:get}}\s{{clipboard}} [output:current file: caret position]

I have more complex workflows for this new variable that I'm thinking about but is not 100% closed in my head (using this new variable to create inline embed content) that will use `Event : After Active Leaf has changed` but I have to think more 

---

Let me know what you think :)

Have a great day!

**Comment by Taitava** [comment]
This is a really nice idea! 😎 

I'd like to make it a bit more general so that it could work regardless of output channel, i.e. output does not have to go to the statusbar. Instead, where ever it goes, it could be captured by some mechanism, and there could be some new variable that can read the previous output.

Just recently, I've had an idea about making it possible do define multiple output channels. So, you could define e.g. output to status bar and to a file at the same time. Now that I'm thinking more about this, we could also add custom variables as an output channel. I don't know yet what naming rules there would be for custom variables as I just invented the idea while typing, but let's say for now that it could be something like this: New output channel named _Custom variable_, which lets a user define a variable name that starts with `{{custom:`. So I could define e.g. `{{custom:command_result}}` as an output variable, and use it when executing another shell command.

Multiple shell commands could write to a same output variable if wanted, and the variable would always give the **latest** value that was written to it. When a shell command writes to a variable, the variable will forget its current value and get a new value. Multiple different variables could be used to avoid overwriting if needed.

So, I came up with a bit different proposal than what you had in mind. Let's work this out so that we can make sure to have a solution for your use case! 🌞 

Thanks and have a nice day, too! 🙂 I'll read your other messages later - maybe tomorrow or after a few days as I have some exams in school that hurt my eager to spend time with this plugin. 😆

**Reply by FelipeRearden** [reply]
> I'd like to make it a bit more general so that it could work regardless of output channel, i.e. output does not have to go to the statusbar. Instead, where ever it goes, it could be captured by some mechanism, and there could be some new variable that can read the previous output.

That's a great idea to define a new framework for this featire @Taitava !!!!

> Just recently, I've had an idea about making it possible do define multiple output channels. So, you could define e.g. output to status bar and to a file at the same time. Now that I'm thinking more about this, we could also add custom variables as an output channel. I don't know yet what naming rules there would be for custom variables as I just invented the idea while typing, but let's say for now that it could be something like this: New output channel named _Custom variable_, which lets a user define a variable name that starts with `{{custom:`. So I could define e.g. `{{custom:command_result}}` as an output variable, and use it when executing another shell command.

> Multiple shell commands could write to a same output variable if wanted, and the variable would always give the **latest** value that was written to it. When a shell command writes to a variable, the variable will forget its current value and get a new value. Multiple different variables could be used to avoid overwriting if needed.

Wooowwww AMAZING @Taitava !!!! this is groundbreaking @Taitava !!!!

> So, I came up with a bit different proposal than what you had in mind. Let's work this out so that we can make sure to have a solution for your use case! 🌞 

This is a great ideia @Taitava . I bring a rock to you and you transform into a 💎 :)

It's a matter of adjusting the FR :)

> Thanks and have a nice day, too! 🙂 I'll read your other messages later - maybe tomorrow or after a few days as I have some exams in school that hurt my eager to spend time with this plugin. 😆 

Sure ! I wish you the best luck in the world on your exams !!!!!!!!!!

I'm very happy to know that you are having fun with SC. It's the best feeling in the world.

We talk soon !!!!!!!!!!!!!!!!!!!!!!

**Reply by Taitava** [reply]
> This is a great ideia @Taitava . I bring a rock to you and you transform into a 💎 :)

Thanks! 🙂 Or: you brought me a 💎 and just shapened it a bit 😉 .

> I'm very happy to know that you are having fun with SC. It's the best feeling in the world.

Yea, it's been nice to program this thing - especially when you are here commenting and giving feedback! 🌞 

Have a nice day! 😎

**Reply by FelipeRearden** [reply]
Thanks @Taitava ! 

I'm glad to know that you like my feedback :) :) :)

Let's keep talking !

**Comment by Taitava** [comment]
# It's finally becoming an output channel

I have created a new output channel called _Assign custom variables_, which will be released in SC `0.22.0` in the near future. A quote from issue #382:

> This output channel assumes the output is a JSON object, reading it's properties and assigning them to custom variables.
>
> For example, consider the following JSON output:
> ```JSON
> {
>     "_link_to_file": "My folder/My note.md",
>     "_link_alias": "Meeting notes"
> }
> ```
>
> This would assign the value `My folder/My note.md` to a custom variable named `{{_link_to_file}}`, and `Meeting notes` to `{{_link_alias}}`.

**Reply by Taitava** [reply]
Just for the record: `0.22.0` was released a while ago.
